Shaders fail to compile with matrix-type global uniforms (4.1 Compatibility renderer)

### Godot version

v4.1.1.stable.official [bd6af8e0e]

### System information

Godot v4.1.1.stable - Linux Mint 20.2 (Uma) - Vulkan (Compatibility) - NVIDIA GeForce GTX 1660 Ti (nvidia; 470.199.02) - Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz (12 Threads)

### Issue description

When using the Compatibility renderer, shaders fail to compile if a matrix-type global uniform is used in `vertex()`. The error given is an internal cast from `int` to `uint`:
```
drivers/gles3/shader_gles3.cpp:242 - SceneShaderGLES3: Vertex shader compilation failed:
error C7011: implicit cast from "int" to "uint"
```

### Steps to reproduce

Add a matrix global shader uniform (like a `mat3`, below) in the project settings. This shader will fail to compile:
```
shader_type spatial;

global uniform mat3 bug;

void vertex() {
	mat3 cannot_compile = bug;
}
```

### Minimal reproduction project

[Reproduction 9-8-23.zip](https://github.com/godotengine/godot/files/12564495/Reproduction.9-8-23.zip)

--- a/servers/rendering/shader_compiler.cpp
+++ b/servers/rendering/shader_compiler.cpp
@@ -432,13 +432,13 @@ static String _get_global_shader_uniform_from_type_and_index(const String &p_buf
 			return "(" + p_buffer + "[" + p_index + "].xyzw)";
 		}
 		case ShaderLanguage::TYPE_MAT2: {
-			return "mat2(" + p_buffer + "[" + p_index + "].xy," + p_buffer + "[" + p_index + "+1].xy)";
+			return "mat2(" + p_buffer + "[" + p_index + "].xy," + p_buffer + "[" + p_index + "+1u].xy)";
 		}
 		case ShaderLanguage::TYPE_MAT3: {
-			return "mat3(" + p_buffer + "[" + p_index + "].xyz," + p_buffer + "[" + p_index + "+1].xyz," + p_buffer + "[" + p_index + "+2].xyz)";
+			return "mat3(" + p_buffer + "[" + p_index + "].xyz," + p_buffer + "[" + p_index + "+1u].xyz," + p_buffer + "[" + p_index + "+2u].xyz)";
 		}
 		case ShaderLanguage::TYPE_MAT4: {
-			return "mat4(" + p_buffer + "[" + p_index + "].xyzw," + p_buffer + "[" + p_index + "+1].xyzw," + p_buffer + "[" + p_index + "+2].xyzw," + p_buffer + "[" + p_index + "+3].xyzw)";
+			return "mat4(" + p_buffer + "[" + p_index + "].xyzw," + p_buffer + "[" + p_index + "+1u].xyzw," + p_buffer + "[" + p_index + "+2u].xyzw," + p_buffer + "[" + p_index + "+3u].xyzw)";
 		}
 		default: {
 			ERR_FAIL_V("void");

